controllers/keystoneendpoint_controller.go

```diff
@@ -0,0 +1,422 @@
+	}
+
+	// Service is deleted so remove the finalizer.
+	controllerutil.RemoveFinalizer(instance, helper.GetFinalizer())
+	util.LogForObject(helper, "Reconciled Endpoint delete successfully", instance)
+
+	if err := r.Update(ctx, instance); err != nil && !k8s_errors.IsNotFound(err) {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *KeystoneEndpointReconciler) reconcileNormal(
+	ctx context.Context,
+	instance *keystonev1.KeystoneEndpoint,
+	helper *helper.Helper,
+	os *openstack.OpenStack,
+) (ctrl.Result, error) {
+	util.LogForObject(helper, "Reconciling Endpoint normal", instance)
+
+	// If the service object doesn't have our finalizer, add it.
+	controllerutil.AddFinalizer(instance, helper.GetFinalizer())
+	// Register the finalizer immediately to avoid orphaning resources on delete
+	if err := r.Update(ctx, instance); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	//
+	// Wait for KeystoneService is Ready and get the ServiceID from the object
+	//
+	ksSvc, err := keystonev1.GetKeystoneServiceWithName(ctx, helper, instance.Spec.ServiceName, instance.Namespace)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	// mirror the Status, Reason, Severity and Message of the latest keystoneservice condition
+	// into a local condition with the type condition.KeystoneServiceReadyCondition
+	c := ksSvc.Status.Conditions.Mirror(condition.KeystoneServiceReadyCondition)
+	if c != nil {
+		instance.Status.Conditions.Set(c)
+	}
+
+	if !ksSvc.IsReady() {
+		util.LogForObject(helper, fmt.Sprintf("KeystoneService %s not ready, waiting to create endpoints", instance.Spec.ServiceName), instance)
+
+		return ctrl.Result{RequeueAfter: time.Duration(10) * time.Second}, nil
+	}
+
+	instance.Status.ServiceID = ksSvc.Status.ServiceID
+
+	//
+	// create/update endpoints
+	//
+	err = r.reconcileEndpoints(
+		instance,
+		helper,
+		os)
+	if err != nil {
+		instance.Status.Conditions.Set(condition.FalseCondition(
+			keystonev1.KeystoneServiceOSEndpointsReadyCondition,
+			condition.ErrorReason,
+			condition.SeverityWarning,
+			keystonev1.KeystoneServiceOSEndpointsReadyErrorMessage,
+			err.Error()))
+		return ctrl.Result{}, err
+	}
+	instance.Status.Conditions.MarkTrue(
+		keystonev1.KeystoneServiceOSEndpointsReadyCondition,
+		keystonev1.KeystoneServiceOSEndpointsReadyMessage,
+		instance.Spec.Endpoints,
+	)
+
+	util.LogForObject(helper, "Reconciled Endpoint normal successfully", instance)
+
+	return ctrl.Result{}, nil
+}
+
+func (r *KeystoneEndpointReconciler) reconcileEndpoints(
+	instance *keystonev1.KeystoneEndpoint,
+	helper *helper.Helper,
+	os *openstack.OpenStack,
+) error {
+	util.LogForObject(helper, "Reconciling Endpoints", instance)
+
+	// check for removed endpoints from the Spec.Endpoints definition
+	if instance.Status.EndpointIDs != nil && len(instance.Spec.Endpoints) < len(instance.Status.EndpointIDs) {
```

Comment:
If the client adds an endpoint and removes another in the same update call then the length of the EndpointIDs will be the same as the length of the Spec.Endpoint but the content will differ. I think in this case this code will not delete the removed endpoint.

Reply:
good call, let me improve that.

Reply:
5817c5c78e2ae547b4b0b7e3c962e1d542c9965b should cover this situation now